File: ejemplos/nerve/04-vigilante-de-carpetas/ejemplo.py
"""
Ejemplo: Monitoreo de carpeta con watchdog

Qué enseña este ejemplo:
  - Cómo usar una librería externa (watchdog) para vigilar el sistema de archivos.
  - Cómo usar manejadores de eventos (event handlers) para reaccionar a cambios.
  - Cómo mantener el programa vivo con un loop (while True) + sleep.

Para tu reto (Vigilante de carpetas con Nerve):
  Aplica esta misma estructura. En vez de solo hacer `print`, cuando se cree un 
  archivo, usarás `cliente.broadcast()` o `cliente.send()` de Nerve para 
  notificar a otros nodos.

Glosario de funciones/librerías:
  FileSystemEventHandler — Una clase base que tiene métodos como on_created, on_modified, etc.
  Observer             — El objeto que se encarga de vigilar la carpeta en un hilo separado.
  time.sleep(1)        — Hace que el programa espere 1 segundo, evitando usar el 100% de la CPU.
"""

import time
import os
from watchdog.observers import Observer
from watchdog.events import FileSystemEventHandler

# Definimos cómo queremos reaccionar a los eventos
class MiManejador(FileSystemEventHandler):
    
    # Se llama automáticamente cuando se crea un archivo o carpeta
    def on_created(self, event):
        # event.src_path contiene la ruta del archivo afectado
        print(f"📁 [NUEVO] Se ha creado: {event.src_path}")
        
    # Se llama cuando se modifica un archivo
    def on_modified(self, event):
        print(f"✏️  [MODIFICADO] Se ha modificado: {event.src_path}")
        
    # Se llama cuando se elimina un archivo
    def on_deleted(self, event):
        print(f"🗑️  [ELIMINADO] Se ha eliminado: {event.src_path}")


def main():
    carpeta_a_vigilar = "./carpeta_prueba"
    
    # Creamos la carpeta si no existe
    if not os.path.exists(carpeta_a_vigilar):
        os.makedirs(carpeta_a_vigilar)
        print(f"Carpeta creada: {carpeta_a_vigilar}")

    # 1. Instanciamos nuestro manejador y el observador
    manejador = MiManejador()
    observador = Observer()
    
    # 2. Le decimos al observador qué carpeta vigilar, y qué manejador usar
    # recursive=False significa que no vigilará subcarpetas.
    observador.schedule(manejador, path=carpeta_a_vigilar, recursive=False)
    
    # 3. Iniciamos el observador (esto ocurre en segundo plano)
    observador.start()
    print(f"👀 Vigilando la carpeta '{carpeta_a_vigilar}'...")
    print("Crea, modifica o elimina archivos ahí para ver los eventos. Presiona Ctrl+C para salir.")

    try:
        # Mantenemos el hilo principal vivo
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        print("\nDeteniendo observador...")
        observador.stop()
        
    # Esperamos a que termine de cerrarse limpiamente
    observador.join()

if __name__ == "__main__":
    main()

File: ejemplos/nerve/04-vigilante-de-carpetas/ejemplo.go
/*
 * Ejemplo: Monitoreo de carpeta con fsnotify en Go
 * 
 * Qué enseña este ejemplo:
 *   - Cómo usar un módulo externo de Go (github.com/fsnotify/fsnotify).
 *   - Cómo usar goroutines (go func) para ejecutar tareas en segundo plano.
 *   - Cómo usar un bucle for-select para leer eventos desde canales (channels).
 * 
 * Para tu reto (Vigilante de carpetas con Nerve):
 *   Aplica esta misma estructura. En vez de solo hacer `log.Println`, cuando se
 *   detecte un evento (ej. Create o Write), usarás el cliente de Nerve para 
 *   notificar a otros nodos.
 * 
 * Glosario de funciones/conceptos:
 *   fsnotify.NewWatcher()  — Crea un nuevo observador del sistema de archivos.
 *   defer watcher.Close()  — Asegura que el observador se cierre al terminar la función (limpieza).
 *   watcher.Add(path)      — Añade una ruta (archivo o carpeta) a la lista de vigilancia.
 *   <-watcher.Events       — Lee (recibe) datos del canal de eventos del observador. Bloquea hasta que haya uno.
 *   select                 — Permite a una goroutine esperar múltiples operaciones de comunicación (canales).
 */

package main

import (
	"log"
	"os"

	"github.com/fsnotify/fsnotify"
)

func main() {
	carpetaAVigilar := "./carpeta_prueba"

	// Creamos la carpeta si no existe (0755 son los permisos estándar de lectura/escritura/ejecución)
	if _, err := os.Stat(carpetaAVigilar); os.IsNotExist(err) {
		err = os.Mkdir(carpetaAVigilar, 0755)
		if err != nil {
			log.Fatal("Error al crear carpeta:", err)
		}
		log.Printf("Carpeta creada: %s\n", carpetaAVigilar)
	}

	// 1. Creamos un nuevo observador (watcher)
	watcher, err := fsnotify.NewWatcher()
	if err != nil {
		log.Fatal(err)
	}
	// defer asegura que los recursos se liberen al terminar main()
	defer watcher.Close()

	// 2. Iniciamos una goroutine (un hilo ligero en segundo plano)
	// que estará escuchando constantemente por eventos o errores.
	go func() {
		for {
			select {
			case event, ok := <-watcher.Events:
				if !ok {
					return
				}
				// Verificamos qué tipo de operación fue (Create, Write, Remove, Rename, Chmod)
				if event.Has(fsnotify.Create) {
					log.Printf("📁 [CREADO] Archivo: %s\n", event.Name)
				}
				if event.Has(fsnotify.Write) {
					log.Printf("✏️  [MODIFICADO] Archivo: %s\n", event.Name)
				}
				if event.Has(fsnotify.Remove) {
					log.Printf("🗑️  [ELIMINADO] Archivo: %s\n", event.Name)
				}
				if event.Has(fsnotify.Rename) {
					log.Printf("🔄 [RENOMBRADO] Archivo: %s\n", event.Name)
				}

			case err, ok := <-watcher.Errors:
				if !ok {
					return
				}
				log.Println("❌ Error:", err)
			}
		}
	}()

	// 3. Añadimos la carpeta que queremos vigilar al watcher
	err = watcher.Add(carpetaAVigilar)
	if err != nil {
		log.Fatal(err)
	}

	log.Printf("👀 Vigilando la carpeta '%s'...", carpetaAVigilar)
	log.Println("Crea, modifica o elimina archivos ahí para ver los eventos. Presiona Ctrl+C para salir.")

	// Bloqueamos el hilo principal para que el programa no termine inmediatamente.
	// <-make(chan struct{}) se queda esperando infinitamente (hasta que lo mates con Ctrl+C).
	<-make(chan struct{})
}
